docs/troubleshoot/app-framework/appserver-error-codes.md

```diff
@@ -2388,283 +2486,332 @@ The following error message codes may appear on the app-server log. Use the foll
 
   **Action:**
 
-  TODO
+  Review the error to resolve it and contact support if needed. It's also possible as a workaround to explicitly state the CAs within the keyring that you would like to load, rather than relying upon the server's attempt to automatically find all CAs within the keyring.
 
 
 
-### ZWED0179W":"Unable to retrieve the list of certificate authorities from the keyring=%s owner=%s Error: %s
+### ZWED0001E
+
+  Error: _error_
 
   **Reason:**
 
-  TODO
+  The server is running in cluster mode and the cluster manager has encountered an unexpected error.
 
   **Action:**
 
-  TODO
+  Review the error to resolve it, and contact support if needed.
 
 
 
-### ZWED0001E":"Error: %s
+### ZWED0002E
+
+  Could not stop language manager for types=_languageName_
 
   **Reason:**
 
-  TODO
+  A plugin had a service that needed a language manager to run. During shutdown, the language manager could not be stopped.
 
   **Action:**
 
-  TODO
+  The language manager may continue to run after the app-server shuts down. Review the logs to determine the location of the language manager and try to stop the manager manually.
 
 
 
-### ZWED0002E":"Could not stop language manager for types=%s
+### ZWED0003E
+
+  Loopback configuration not valid, _loobackConfiguration_
+  Loopback calls will fail!
 
   **Reason:**
 
-  TODO
+  The loopback configuration that the server uses to contact itself over an internal network was missing a value for the network port, therefore no requests over the loopback address will be possible.
 
   **Action:**
 
-  TODO
+  Review the configuration of components.app-server.node.port to see if it has a value and set one to fix the issue.
+
 
 
+### ZWED0004E
 
-### ZWED0003E":"Loopback configuration not valid,%s\nLoopback calls will fail!
+  Could not listen on address _ip_:_port_. It is already in use by another process.
 
   **Reason:**
 
-  TODO
+  The server tried to start using the ip and port values shown which were from the zowe configuration. When trying to connect to this address, the server recieved an error telling it that the address was already in use.
 
   **Action:**
 
-  TODO
+  Check the system's network port status to see what program could be using this address, and either stop that program or change the zowe configuration to use a different address before restarting zowe.
+
 
 
+### ZWED0005E
 
-### ZWED0004E":"Could not listen on address %s:%s. It is already in use by another process.
+  Could not listen on address _ip_:_port_. Invalid IP for this system.
 
   **Reason:**
 
-  TODO
+  When the app-server was beinging to the address shown, it recieved the error EADDRNOTAVAIL or ENOTFOUND. In either case, the app-server was not able to bind to the address and so it will not run until the problem is solved.
 
   **Action:**
 
-  TODO
+  Review the address and check if there if it is valid or if there is some lack of permissions that might explain why these errors were received by the server.
```

Comment:
Review the address and **check if it is valid** or ...

Reply:
thank you